src/compiler/checker.ts

```diff
@@ -28417,6 +28417,47 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkObjectLiteral(node: ObjectLiteralExpression, checkMode?: CheckMode): Type {
+        const contextualType = getContextualType(node, /*contextFlags*/ undefined);
+        const isContextualTypeDependent =
+            contextualType &&
+            contextualType.immediateBaseConstraint &&
+            contextualType.immediateBaseConstraint.aliasTypeArguments &&
+            contextualType.immediateBaseConstraint.aliasTypeArguments.length === 1 &&
+            contextualType.immediateBaseConstraint.aliasTypeArguments[0].id === contextualType.id;

```

Comment:
How do you check if type A is depdendent on type B? That is I want to do `doesTypeDependOn(contextualType.immediateBaseConstraint, contextualType)`.

Reply:
Perhaps you want to use [`isTypeParameterPossiblyReferenced`](https://github.dev/microsoft/TypeScript/blob/9d516b64ca6c7ce91fa21a2e8c5d62268e47cf86/src/compiler/checker.ts#L17481) or a similar thing?

Reply:
Actually I realised not only it would be expensive but also I don't need a function that lookups the type tree of A and finds if it has B, I want a function that looks up the ast tree of A and see if there is a reference to B. So I want something roughly like...

```ts
const isContextualTypeDependent =
  contextualType &&
  (contextualType & TypeFlags.TypeParameter) &&
  !!forEachChildRecursively(contextualType.symbol.declarations[0].constraint, node => {
    if (
      node.kind === SyntaxKind.TypeReference &&
      getSymbolOfNode(node.typeName).id ===
      getSymbolOfNode(contextualType.symbol.declarations[0]).id
    ) {
      return true
    }
  })
```